Authorization/Authorization/Presentation/Login/SignInViewModel.swift

```diff
@@ -29,20 +50,78 @@ public class SignInViewModel: ObservableObject {
             }
         }
     }
+    var oauthswift: OAuth2Swift?
     
     private let interactor: AuthInteractorProtocol
     let router: AuthorizationRouter
+    let config: Config
     let analytics: AuthorizationAnalytics
     private let validator: Validator
+    private var safariDelegate: WebLoginSafariDelegate?
     
     public init(interactor: AuthInteractorProtocol,
                 router: AuthorizationRouter,
                 analytics: AuthorizationAnalytics,
+                config: Config,
                 validator: Validator) {
         self.interactor = interactor
         self.router = router
         self.analytics = analytics
+        self.config = config
         self.validator = validator
+        self.webLoginAttempted = false
+    }
+    
+    @MainActor
+    func login(viewController: UIViewController) async {
+        /* OAuth web login. Used when we cannot use the built-in login form,
+        but need to let the LMS redirect us to the authentication provider.
+        
+        An example service where this is needed is something like Auth0, which
+        redirects from the LMS to its own login page. That login page then redirects
+        back to the LMS for the issuance of a token that can be used for making
+        requests to the LMS, and then back to the redirect URL for the app. */
+        self.safariDelegate = WebLoginSafariDelegate(viewModel: self)
+        oauthswift = OAuth2Swift(
+            consumerKey: config.oAuthClientId,
+            consumerSecret: "", // No secret required
+            authorizeUrl: "\(config.baseURL)/oauth2/authorize/",
+            accessTokenUrl: "\(config.baseURL)/oauth2/access_token/",
+            responseType: "code"
+        )
+        
+        oauthswift!.allowMissingStateCheck = true
+        let handler = SafariURLHandler(
+            viewController: viewController, oauthSwift: oauthswift!
+        )
+        handler.delegate = self.safariDelegate
+        oauthswift!.authorizeURLHandler = handler
+
+        // Trigger OAuth2 dance
+        guard let rwURL = URL(string: "\(Bundle.main.bundleIdentifier ?? "")://oauth2Callback") else { return }
+        oauthswift!.authorize(withCallbackURL: rwURL, scope: "", state: "") { result in
+            switch result {
+            case .success(let (credential, _, _)):
+                Task {
+                    self.webLoginAttempted = true
+                    let user = try await self.interactor.login(credential: credential)
+                    self.analytics.setUserID("\(user.id)")
+                    self.analytics.userLogin(method: .oauth2)
+                    self.router.showMainScreen()
+                }
+                // Do your request
+            case .failure(let error):
+                self.webLoginAttempted = true
+                self.isShowProgress = false
+                self.errorMessage = error.localizedDescription
+            }
+        }
+    }
```

Comment:
This function can be problematic for several reasons.
1. We are trying to encapsulate all navigation and ViewControllers logic inside of the Router entity.
This helps us to free ourselves from UIKit, navigation logic and be able to change the navigation framework at any time and write tests easily.
We could play with it and ask Router to provide us with configured SafariURLHandler instance or something like that.
2. I recommend converting the closure style function
`oauthswift!.authorize(withCallbackURL: rwURL, scope: "", state: "") { result in`
to async style function
`await oauthswift!.authorize(...)`  or `await oauthswift!.authorizeTask(...)`
we can reach it with some extension function and wrap `oauthswift!.authorize` to the [withCheckedThrowingContinuation](https://developer.apple.com/documentation/swift/withcheckedthrowingcontinuation(function:_:)) call.
3. It's better to remove `Task {}` creating from the async function as it makes it hard/impossible to test with Unit testing.
4. Last but not least, could you please extend the `SignInViewModelTests.swift` with tests for your new functions? This is a crucial part of the application's stability.
I highly recommend running the `generateAllMocks.sh` script before you start creating tests. This script creates all mocks for tests if its need.

If you need any help, hit me on Slack or email me 📪

Reply:
Thanks, @volodymyr-chekyrta ! This is great feedback. I'll ping you as I make progress on it, as this appears to be the bulk of the changes to be made.